### PR TITLE
feat(cli): support multi-line status line output

### LINKED ISSUE #3211: feat: Support multi-line status line output
## Is your feature request related to a problem?

The current status line is hardcoded to only use the first line of script output:

```ts
const trimmed = stdout.trim().split('\n')[0] || '';
```

This prevents users from displaying richer status information, such as:
- Context usage with a visual progress bar
- System metrics (CPU, memory) on separate rows
- Rate limit countdown timers
- Environment context (Node/Python versions)

## Describe the solution you'd like

1. **Remove the `split('\n')[0]` hard limit** — allow the full stdout to be used
2. **Track output line count** — count how many lines the script produces
3. **Clear before re-render** — use ANSI escape sequences to erase old lines before writing new ones:
   ```
   \x1b[N A   # cursor up N lines
   \x1b[J     # erase from cursor to end of screen
   ```
4. **Keep horizontal truncation** — `wrap="truncate"` should remain for long single lines

### Reference implementation

This approach is already used in Claude Code's status line rendering (`packages/core/src/bridge/bridgeUI.ts`):

```ts
let statusLineCount = 0

function clearStatusLines() {
  if (statusLineCount <= 0) return
  write(`\x1b[${statusLineCount}A`)  // cursor up N lines
  write('\x1b[J')                     // erase to end of screen
  statusLineCount = 0
}

function writeStatus(text: string) {
  write(text)
  statusLineCount += countVisualLines(text)
}
```

### Suggested approach for Qwen Code

```tsx
const [lineCount, setLineCount] = useState(0)

function renderStatusLine(output: string) {
  if (lineCount > 0) {
    process.stdout.write(`\x1b[${lineCount}A`)
    process.stdout.write('\x1b[J')
  }

  const lines = output.split('\n').filter(Boolean)
  setLineCount(lines.length)
  lines.forEach(line => process.stdout.write(line + '\n'))
}
```

## Additional context

The single-line limitation was added during the original PR review as a layout safeguard. Multi-line support would make the feature more flexible without breaking existing single-line scripts.

**Willing to submit a PR?** Yes, if maintainers are open to this direction.

### PR DESCRIPTION
**Pull request by wenshao**
## Summary

Closes #3211

Remove the single-line hard limit (`.split('\n')[0]`) from `useStatusLine` so user scripts can output up to 2 lines. The hook now returns `lines: string[]` directly — Footer maps over the array to render each line as a `<Text wrap="truncate">` element. Empty lines are filtered so leading blanks don't consume the line cap. Ink's virtual DOM handles re-rendering automatically — no manual ANSI cursor management needed.

### Before / After

When a status line is active, the "? for shortcuts" hint is suppressed. The status line command output occupies the left side of the footer.

```
Before (single-line only — 1 row):
┌─────────────────────────────────────────────────────────────────┐
│  qwen-3-235b (main) ctx:34%          🔒 docker | Debug | 67%   │
└─────────────────────────────────────────────────────────────────┘

After (up to 2 status lines — 2 rows):
┌─────────────────────────────────────────────────────────────────┐
│  qwen-3-235b (main) ctx:34%          🔒 docker | Debug | 67%   │
│  ████████░░░░░░░░░░ 34% context                                │
└─────────────────────────────────────────────────────────────────┘

After + non-default approval mode (3 rows max):
┌─────────────────────────────────────────────────────────────────┐
│  qwen-3-235b (main) ctx:34%          🔒 docker | Debug | 67%   │
│  ████████░░░░░░░░░░ 34% context                                │
│  auto-accept edits (shift + tab to cycle)                       │
└─────────────────────────────────────────────────────────────────┘
```

### Backward compatibility

Single-line scripts produce identical output — the only behavioral change is that a second line is now kept instead of discarded.

### Output limits

- **Line cap**: \`MAX_STATUS_LINES = 2\` (exported, with comment explaining the 3-row footer constraint) — lines beyond the second are silently discarded
- **Empty-line filter**: \`.filter(Boolean)\` strips blank lines before capping, so e.g. \`"\\n\\nreal content"\` yields \`["real content"]\` instead of \`["", ""]\`
- **Buffer cap**: \`maxBuffer: 10KB\` (existing) — scripts exceeding this trigger an exec error and the status line clears

## Changes

- **\`useStatusLine.ts\`** — return \`lines: string[]\` (was \`text: string | null\`); strip trailing newline, split, filter empty lines, cap at \`MAX_STATUS_LINES\`
- **\`Footer.tsx\`** — consume \`lines\` array directly (no re-split); render each line as \`<Text key={\\\`status-line-\${i}\\\`} dimColor wrap="truncate">\`
- **\`builtin-agents.ts\`** — update statusline setup agent prompt to mention multi-line support (up to 2 lines)
- **\`useStatusLine.test.ts\`** — 40 tests (was 31): multi-line, 2-line cap, empty-line filter, \\r\\n, pure-newline, metrics/fallback coverage
- **\`Footer.test.tsx\`** — 8 tests (was 6): add multi-line rendering and hint suppression tests
- **\`status-line.md\`** — update ASCII diagrams (3 scenarios), field description, and behavior section

## Test plan

- [x] \`useStatusLine.test.ts\` — 40 tests pass (100% line coverage)
- [x] \`Footer.test.tsx\` — 8 tests pass, snapshots unchanged
- [x] \`statuslineCommand.test.ts\` — 4 tests pass
- [x] TypeScript — clean (our files; unrelated hooks PR has pre-existing errors)
- [x] ESLint — clean on all changed files
- [x] Manual: configure a multi-line statusline script, verify each line renders with truncation

### CODE REVIEW

File: packages/cli/src/ui/hooks/useStatusLine.ts [hunk 1]
```diff
@@ -277,8 +278,9 @@ export function useStatusLine(): {
         activeChildRef.current = undefined;
         if (!error && stdout) {
           // Strip only the trailing newline to preserve intentional whitespace.
-          const line = stdout.replace(/\r?\n$/, '').split(/\r?\n/, 1)[0];
-          setOutput(line || null);
+          const lines = stdout.replace(/\r?\n$/, '').split(/\r?\n/);
+          const text = lines.slice(0, MAX_STATUS_LINES).join('\n');
```
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.ts]: **[suggestion]** hook 内部用 `join('\n')` 拼回字符串，Footer 里又用 `split(/\r?\n/)` 拆开——等于序列化了一个数组再立即反序列化。

建议 hook 直接返回 `string[]`（例如 `lines: string[]`），Footer 直接 `.map` 渲染，避免这层无意义的 join/split 往返。对外 API 也更清晰：调用方拿到的是结构化数据而非需要再次解析的字符串。

File: packages/cli/src/ui/hooks/useStatusLine.ts [hunk 2]
```diff
@@ -277,8 +278,9 @@ export function useStatusLine(): {
         activeChildRef.current = undefined;
         if (!error && stdout) {
           // Strip only the trailing newline to preserve intentional whitespace.
-          const line = stdout.replace(/\r?\n$/, '').split(/\r?\n/, 1)[0];
-          setOutput(line || null);
+          const lines = stdout.replace(/\r?\n$/, '').split(/\r?\n/);
```
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.ts]: **[bug]** 当脚本输出 `"\n\nreal content\n"` 时，strip trailing newline 后得到 `"\n\nreal content"`，split 得到 `["", "", "real content"]`，`slice(0, 2)` 取到 `["", ""]`——真正有内容的那行被丢弃了，用户看到的是两行空白。

建议在 slice 之前先过滤掉空行（或者至少过滤掉前导空行）：
```ts
const lines = stdout.replace(/\r?\n$/, '').split(/\r?\n/).filter(l => l.length > 0);
```
这样也能保持 `MAX_STATUS_LINES` 的语义——"最多 2 行**有效内容**"。

File: packages/cli/src/ui/hooks/useStatusLine.ts [hunk 3]
```diff
@@ -72,6 +72,7 @@ interface StatusLineConfig {
 }
 
 const debugLog = createDebugLogger('STATUS_LINE');
+const MAX_STATUS_LINES = 2;
```
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.ts]: **[suggestion]** `MAX_STATUS_LINES` 定义在 hook 里，但 Footer 渲染时并不引用这个常量——它只是盲目地渲染 hook 返回的所有行。如果将来有人把 `MAX_STATUS_LINES` 改成 3，Footer 会自动多渲染一行，但没有人会意识到需要验证 footer 高度是否还合理。

建议：
1. export 这个常量，Footer 里加一个防御性注释引用它；
2. 或者在注释里明确说明 footer 总高度约束（status lines + mode indicator <= 3 rows）与这个常量的关系。

File: packages/cli/src/ui/hooks/useStatusLine.test.ts [hunk 4]
```diff
@@ -211,15 +211,37 @@ describe('useStatusLine', () => {
...
+    it('preserves empty lines in multi-line output', async () => {
+      setStatusLineConfig({ type: 'command', command: 'echo lines' });
+      const { result } = renderHook(() => useStatusLine());
+
+      await act(async () => {
+        execCallback(null, '\nline two\n', '');
+      });
+
+      expect(result.current.text).toBe('\nline two');
+    });
```
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.test.ts:234]: **[nit]** 这个测试验证了空行会被保留（`'\nline two'`），但结合 Footer 的渲染逻辑，这意味着用户会看到一行空白 + 一行内容，体验上比较奇怪。

如果采纳了上面过滤空行的建议，这个测试的预期值应改为 `'line two'`。如果确实想保留空行作为有意设计，建议在测试注释里说明理由。

File: packages/cli/src/ui/components/Footer.tsx [hunk 5]
```diff
@@ -120,13 +122,14 @@ export const Footer: React.FC = () => {
...
       <Box flexDirection="column" flexShrink={isNarrow ? 0 : 1}>
-        {statusLineText &&
+        {statusLineLines.length > 0 &&
           !uiState.ctrlCPressedOnce &&
-          !uiState.ctrlDPressedOnce && (
-            <Text dimColor wrap="truncate">
-              {statusLineText}
+          !uiState.ctrlDPressedOnce &&
+          statusLineLines.map((line, i) => (
+            <Text key={i} dimColor wrap="truncate">
```
Comment (BZ-D) [inline packages/cli/src/ui/components/Footer.tsx]: **[nit]** `key={i}` 对于一个最多 2 个元素且总是整体重渲染的列表来说问题不大。但如果后续行数/顺序可能变化，index key 会导致 React 复用错误的 DOM 节点。可以考虑用 `` key={`status-line-${i}`} `` 做 key，更明确一些。

**Review by BZ-D — COMMENTED** [review]
整体方案合理，多行状态栏是个实用的功能增强。以下是一些具体的代码层面建议。

File: packages/cli/src/ui/components/Footer.tsx [hunk 6]
```diff
@@ -120,13 +122,14 @@ export const Footer: React.FC = () => {
     >
       {/* Left column — status line on top, hints/mode on bottom */}
       <Box flexDirection="column" flexShrink={isNarrow ? 0 : 1}>
-        {statusLineText &&
+        {statusLineLines.length > 0 &&
```
Comment (BZ-D) [inline packages/cli/src/ui/components/Footer.tsx:149]: **[testing gap]** `Footer.test.tsx` 里没有任何测试覆盖 status line 的渲染（现有测试的 mock settings 里没有 `statusLine` 配置，所以 `statusLineText` 始终为 `null`）。

这次 PR 把 Footer 的 status line 渲染逻辑从单个 `<Text>` 改成了 `statusLineLines.map(...)`，但没有对应的 Footer 测试来验证多行渲染是否正确。建议至少补一个 Footer 测试：mock `useStatusLine` 返回含 `\n` 的字符串，断言 `lastFrame()` 中出现两行内容。

File: packages/cli/src/ui/hooks/useStatusLine.test.ts [hunk 7]
```diff
@@ -211,15 +211,37 @@ describe('useStatusLine', () => {
...
+      const { result } = renderHook(() => useStatusLine());
+
+      await act(async () => {
+        execCallback(null, 'line1\nline2\nline3\nline4\n', '');
+      });
+
+      expect(result.current.text).toBe('line1\nline2');
     });
 
     it('returns null when command fails', async () => {
```
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.test.ts:269]: **[testing gap]** 缺少 `\r\n`（Windows 风格换行）的测试。代码里用 `split(/\r?\n/)` 显式处理了 `\r\n`，但没有对应的测试覆盖。建议加一个：
```ts
it('handles \\r\\n line endings', async () => {
  setStatusLineConfig({ type: 'command', command: 'echo lines' });
  const { result } = renderHook(() => useStatusLine());
  await act(async () => {
    execCallback(null, 'line1\r\nline2\r\n', '');
  });
  expect(result.current.text).toBe('line1\nline2');
});
```

File: packages/cli/src/ui/hooks/useStatusLine.ts (hunk 2 above)
Comment (BZ-D) [inline packages/cli/src/ui/hooks/useStatusLine.ts]: **[edge case]** 补充上面空行 bug 的另一个场景：当 stdout 为纯换行（如 `'\n\n'`）时，strip trailing → `'\n'`，split → `['', '']`，join → `'\n'`，这是 truthy 值，会 `setOutput('\n')`。Footer 会渲染 2 行空白占据 footer 空间。

如果采纳了之前的 `.filter(l => l.length > 0)` 建议，这个场景也会被自动修复（filter 后为空数组 → join 为 `''` → `setOutput(null)`）。

**Review by BZ-D — COMMENTED** [review]
补充几个额外发现的问题。

File: packages/cli/src/ui/hooks/useStatusLine.ts (hunk 1 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.ts]: Adopted — hook now returns `lines: string[]` directly, Footer maps over it without any split. The join/split round-trip is eliminated.

File: packages/cli/src/ui/hooks/useStatusLine.ts (hunk 2 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.ts]: Good catch — fixed. Added `.filter(Boolean)` before `.slice()` so empty lines are stripped first. `"\n\nreal content"` now correctly yields `["real content"]` instead of `["", ""]`.

File: packages/cli/src/ui/hooks/useStatusLine.ts (hunk 3 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.ts]: Adopted — `MAX_STATUS_LINES` is now exported with a comment explaining the footer 3-row constraint.

File: packages/cli/src/ui/hooks/useStatusLine.test.ts (hunk 4 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.test.ts:234]: Updated — test renamed to "filters empty lines from output", verifies `"\n\nreal content\n"` yields `["real content"]`.

File: packages/cli/src/ui/components/Footer.tsx (hunk 5 above)
Comment (wenshao) [reply packages/cli/src/ui/components/Footer.tsx]: Adopted — changed to `key={`status-line-${i}`}`.

File: packages/cli/src/ui/components/Footer.tsx (hunk 6 above)
Comment (wenshao) [reply packages/cli/src/ui/components/Footer.tsx:149]: Added — Footer test now mocks `useStatusLine` to return multi-line data and verifies both lines render. Also added a test confirming hint suppression when status line is active.

File: packages/cli/src/ui/hooks/useStatusLine.test.ts (hunk 7 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.test.ts:269]: Added — new test `handles \\r\\n line endings` verifies `"line1\\r\\nline2\\r\\n"` yields `["line1", "line2"]`.

File: packages/cli/src/ui/hooks/useStatusLine.ts (hunk 2 above)
Comment (wenshao) [reply packages/cli/src/ui/hooks/useStatusLine.ts]: Already fixed by the `.filter(Boolean)` change — `"\\n\\n"` now yields `[]` (empty array). Added a dedicated test `returns empty when stdout is only newlines` to cover this edge case.

**Comment by BZ-D** [comment]
There are some merge conflicts, please resolve them first.

**Comment by BZ-D** [comment]
## Bug met in e2e test

When showing multiple lines of status, the right section of Footer aligns to center, as it shows below, just like the content hangs there.

image

Suggestion:

- Add `alignItems='flex-start'` to the right section Box in Footer.tsx, which will align the right content to the top of Footer.

**Comment by wenshao** [comment]
Thanks for the e2e catch! Fixed in ec8f45da9 — added `alignItems="flex-start"` to the right section Box so indicators anchor to the top instead of centering when the left column grows taller.

Also, merge conflict was resolved earlier in 7f177576a (main removed `compactMode` from the footer via #3100; kept our `lines` API).

**Comment by wenshao** [comment]
Also merged latest main again in 5136c882f — resolved new conflicts from #3087 (auto-memory/dream) and #3264 (statusline sync-exec crash fix):

- `Footer.tsx` — kept our `lines` API alongside main's new `useDreamRunning` hook
- `useStatusLine.ts` — nested our filter+cap logic inside main's try/catch around `exec()`, converted `setOutput(null)` on sync-throw to `setOutput([])`
- tests updated to use the `lines` array API

Branch is now conflict-free against origin/main. 50 tests pass.

**Comment by BZ-D** [comment]
E2E Test Passed.

✅ Normal Case:

image

✅ Narrow Case:

image

**Review by BZ-D — APPROVED** [review]
LGTM
